lib/Onyx.js

```diff
@@ -1025,17 +1028,37 @@ function initializeWithDefaultKeyStates() {
  * Storage.setItem() from Onyx.clear() will have already finished and the merged
  * value will be saved to storage after the default value.
  *
+ * @param {Array} keysToPreserve is a list of ONYXKEYS that should not be cleared with the rest of the data
  * @returns {Promise<void>}
  */
-function clear() {
+function clear(keysToPreserve = []) {
     return getAllKeys()
         .then((keys) => {
-            _.each(keys, (key) => {
-                const resetValue = lodashGet(defaultKeyStates, key, null);
-                cache.set(key, resetValue);
-                notifySubscribersOnNextTick(key, resetValue);
+            const keyValuesToClear = [];
+            const defaultKeys = _.keys(defaultKeyStates);
+
+            // The only keys that should be cleared are:
+            // 1. Anything specifically passed in keysToPreserve (because some keys like language preferences, offline status, or activeClients need to remain in Onyx even when signed out)
+            // 2. Any keys with a default state (because they need to remain in Onyx as their default, and setting them to null would cause unknown behavior)
+            const keysToClear = _.difference(keys, keysToPreserve, defaultKeys);
+            keyValuesToClear.push(..._.map(keysToClear, key => [key, null]));
+
+            // Remove any keysToPreserve from the defaultKeyStates because if they are passed in it has been explicitly called out to preserve those values instead of resetting them back
+            // to the default.
+            const defaultKeyValuePairs = _.pairs(_.omit(defaultKeyStates, ...keysToPreserve));
+
+            // Add the default key value pairs to the keyValuesToClear so that they get set back to their default values when we clear Onyx
+            keyValuesToClear.push(...defaultKeyValuePairs);
+
+            // Make sure that we also reset the cache values after clearing the values from storage.
```

Comment:
Here you say that we reset the cache values after clearing from storage and on the next line you say the opposite?